res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3066 @@
+
+
+
+// The LEDs only support 16 base colors. Adding 1 in addition to
+// the normal 2 for Button.prototype.brightnessOn changes the color
+// slightly, so use that get 25 different colors to include the Filter
+// button as a 5th effect chain preset selector.
+const QuickEffectPresetColors = [
+    LedColors.red,
+    LedColors.blue,
+    LedColors.yellow,
+    LedColors.purple,
+    LedColors.white,
+
+    LedColors.magenta,
+    LedColors.azalea,
+    LedColors.salmon,
+    LedColors.red + 1,
+
+    LedColors.sky,
+    LedColors.celeste,
+    LedColors.fuscia,
+    LedColors.blue + 1,
+
+    LedColors.carrot,
+    LedColors.orange,
+    LedColors.honey,
+    LedColors.yellow + 1,
+
+    LedColors.lime,
+    LedColors.aqua,
+    LedColors.green,
+    LedColors.purple + 1,
+
+    LedColors.magenta + 1,
+    LedColors.azalea + 1,
+    LedColors.salmon + 1,
+    LedColors.fuscia + 1,
+];
+
+// assign samplers to the crossfader on startup
+const SamplerCrossfaderAssign = true;
+
+const MotorWindUpMilliseconds = 1200;
+const MotorWindDownMilliseconds = 900;
+
+/*
+ * HID packet parsing library
+ */
+class HIDInputPacket {
+    constructor(reportId) {
+        this.reportId = reportId;
+        this.fields = [];
+    }
+
+    registerCallback(callback, byteOffset, bitOffset, bitLength, defaultOldData) {
+        if (typeof callback !== "function") {
+            throw Error("callback must be a function");
+        }
+
+        if (byteOffset === undefined || typeof byteOffset !== "number" || !Number.isInteger(byteOffset)) {
+            throw Error("byteOffset must be 0 or a positive integer");
+        }
+
+        if (bitOffset === undefined) {
+            bitOffset = 0;
+        }
+        if (typeof bitOffset !== "number" || bitOffset < 0 || !Number.isInteger(bitOffset)) {
+            throw Error("bitOffset must be 0 or a positive integer");
+        }
+
+        if (bitLength === undefined) {
+            bitLength = 1;
+        }
+        if (typeof bitLength !== "number" || bitLength < 1 || !Number.isInteger(bitOffset) || bitLength > 32) {
+            throw Error("bitLength must be an integer between 1 and 32");
+        }
+
+        const field = {
+            callback: callback,
+            byteOffset: byteOffset,
+            bitOffset: bitOffset,
+            bitLength: bitLength,
+            oldData: defaultOldData
+        };
+        this.fields.push(field);
+
+        return {
+            disconnect: () => {
+                this.fields = this.fields.filter((element) => {
+                    return element !== field;
+                });
+            }
+        };
+    }
+
+    handleInput(byteArray, bufferHasNoReportID) {
+        const offset = bufferHasNoReportID ? -1 : 0;
+        const view = new DataView(byteArray);
+        if (!bufferHasNoReportID && view.getUint8(0) !== this.reportId) {
+            return;
+        }
```

Comment:
The offset was a workaround as mentioned above. I will remove it if the spread operator works/I can insert a report ID on manually fetched report.
The impossible case is a safeguard to make sure the the wrong handler hasn't been called, since there could still an accidentally wrong report mapping if someone was adding support for more reports. This was code written by @Be-ing which I think make sense, but if you feel strongly about it, I can remove it